ui/vty/src/Ariadne/UI/Vty/Widget.hs

```diff
@@ -0,0 +1,394 @@
+import IiExtras (postfixLFields)
+
+import qualified Brick as B
+import qualified Brick.Focus as B
+import qualified Data.Map.Strict as Map
+
+import Ariadne.UI.Vty.Face
+import Ariadne.UI.Vty.Keyboard
+import Ariadne.UI.Vty.Scrolling
+
+----------------------------------------------------------------------------
+-- Types and instances
+----------------------------------------------------------------------------
+
+data WidgetNamePart
+  = WidgetNameSelf  -- Special name for adding widget itself to focus list
+
+  | WidgetNameMenu
+  | WidgetNameTree
+  | WidgetNameStatus
+  | WidgetNameHelp
+  | WidgetNameAbout
+  | WidgetNameLogs
+
+  | WidgetNameRepl
+  | WidgetNameReplInput
+
+  | WidgetNameAddWallet
+  | WidgetNameAddWalletNewName
+  | WidgetNameAddWalletNewPass
+  | WidgetNameAddWalletNewButton
+  | WidgetNameAddWalletRestoreName
+  | WidgetNameAddWalletRestoreMnemonic
+  | WidgetNameAddWalletRestorePass
+  | WidgetNameAddWalletRestoreFull
+  | WidgetNameAddWalletRestoreButton
+
+  | WidgetNameWallet
+  | WidgetNameWalletSendAddress
+  | WidgetNameWalletSendAmount
+  | WidgetNameWalletSendPass
+  | WidgetNameWalletSendButton
+
+  | WidgetNameAccount
+  deriving (Eq, Ord, Show)
+
+data WidgetEvent
+  = WidgetEventMenuSelected
+  | WidgetEventButtonPressed
+  | WidgetEventEditChanged
+
+type WidgetName = [WidgetNamePart]
+
+data WidgetEventResult
+  = WidgetEventHandled
+  | WidgetEventNotHandled
+
+data WidgetInfo s p = WidgetInfo
+  { widgetName :: !WidgetName
+  , widgetState :: s
+  , widgetChildren :: !(Map WidgetNamePart (Widget (WidgetInfo s p)))
+  , widgetFocusList :: ![WidgetNamePart]
+  , widgetEventHandlers :: !(Map WidgetNamePart (WidgetEvent -> WidgetEventM s p ()))
+  , widgetEventQueue :: ![(WidgetNamePart, WidgetEvent)]
+  , widgetEventSend :: WidgetEvent -> WidgetEventM s p ()
+  , widgetDraw :: !(s -> WidgetDrawM s p (B.Widget WidgetName))
+  , widgetDrawWithFocused :: !(Bool -> s -> WidgetDrawM s p (B.Widget WidgetName))
+  , widgetDrawWithFocus :: !(WidgetName -> s -> WidgetDrawM s p (B.Widget WidgetName))
+  , widgetHandleEditKey :: !(KeyboardEditEvent -> WidgetEventM s p WidgetEventResult)
+  , widgetHandleKey :: !(KeyboardEvent -> WidgetEventM s p WidgetEventResult)
+  , widgetHandlePaste :: !(Text -> WidgetEventM s p WidgetEventResult)
+  , widgetHandleMouseDown :: !(B.Location -> WidgetEventM s p WidgetEventResult)
+  , widgetHandleScroll :: !(ScrollingAction -> WidgetEventM s p WidgetEventResult)
+  , widgetHandleEvent :: !(UiEvent -> WidgetEventM s p ())
+  }
+
+instance B.Named (WidgetInfo s p) WidgetName where
+  getName = widgetName
+
+data Widget p = forall s. Widget (WidgetInfo s p)
+
+instance B.Named (Widget p) WidgetName where
+  getName (Widget WidgetInfo{..}) = widgetName
+
+type WidgetInitM s p = State (WidgetInfo s p) ()
+
+type WidgetDrawM s p a = ReaderT (WidgetInfo s p) (Reader p) a
+
+type WidgetEventM s p a = StateT s (StateT (WidgetInfo s p) (StateT p (B.EventM WidgetName))) a
```

Comment:
So once again, why are nested `StateT`s and `ReaderT`s (above) necessary?

Reply:
Because I want three things to be available:
* widget-local state `s`
* widget info `WidgetInfo s p`
* parent widget `p`
Moreover, I want lenses for `s` to just work.
The other things could be merged into a tuple or some other structure, but it would require more API methods to access them.

Reply:
Did you consider using classy lenses? https://michaelxavier.net/posts/2016-04-03-Enterprise-Haskell-Pattern-Lensed-Reader.html

Reply:
I didn't, as I didn't stumble on this article.
Looks promising, but should probably be left for a next round of refactoring, which will be much less invasive and more mechanical.